File: acwing/蓝桥杯每日一题2025/week2/6131.农夫约翰最喜欢的操作/6131.农夫约翰最喜欢的操作.cpp
#include <iostream>
#include <algorithm>
#include <cstring>
#include <cstdio>

using namespace std;

const int N = 2 * 1e5 + 10;
typedef long long LL;

int w[N * 2];
LL s[2 * N];

int main()
{
    int t;
    scanf("%d", &t);
    while (t--)
    {
        int n, m;
        scanf("%d%d", &n, &m);
        for (int i = 1; i <= n; i++)
        {
            scanf("%d", &w[i]);
            w[i] %= m;
        }
        sort(w + 1, w + 1 + n);
        for (int i = 1; i <= n; i++)
            w[i + n] = w[i] + m;
        for (int i = 1; i <= 2 * n; i++)
            s[i] = s[i - 1] + w[i];
        LL res = 1e18;
        for (int i = 1; i <= n; i++)
        {
            int l = i, r = l + n - 1;
            int p = (l + r) / 2;
            LL left = (p - l + 1) * (LL)w[p] - (s[p] - s[l - 1]);
            LL right = s[r] - s[p] - (r - p) * (LL)w[p];
            res = min(res, left + right);
        }
        printf("%lld\n", res);
    }
    return 0;
}

File: acwing/蓝桥杯每日一题2025/week2/6131.农夫约翰最喜欢的操作/6131.农夫约翰最喜欢的操作.py
N = 2 * int(1e5) + 10

s = [0] * 2 * N

t = int(input())

for _ in range(t):
    n, m = map(int, input().split())
    w = [-1] + list(map(int, input().split()))

    for i in range(1, n + 1):
        w[i] = w[i] % m

    w.sort()

    for i in range(1, n + 1):
        w.append(w[i] + m)

    for i in range(1, 2 * n + 1):
        s[i] = s[i - 1] + w[i]

    res = float("inf")
    for l in range(1, n + 1):
        r = l + n - 1
        p = (l + r) // 2
        left = (p - l + 1) * w[p] - (s[p] - s[l - 1])
        right = s[r] - s[p] - (r - p) * w[p]
        res = min(res, left + right)

    print(res)
